README.md
# Badges
This application allows you to generate and cache badges for all sorts of statuses you may want for your `README.md`.  It allows you to update the badge while maintaining a static URL for your `README.md`.

# Endpoints

## GET /badges
Get a list of all badges

## GET /badges/< owner >
Get a list of all badges for a specific owner

## GET /badges/< owner >/< project >/< name >
Get details on a specific badge

## GET /badges/< owner >/< project >/< name >/badge.svg
Get SVG image of a badge

## POST /badges/< owner >/< project >/< name >
Create a new badge

### JSON Data
* subject - (defaults to `name`) What you want to track the value/status for
* color - The color of the badge
* status - The value/status of the subject

Visit https://shields.io towards the bottom at **Your Badge** for an idea of configuration options.

Sample post data:
```
{
  "color":"green",
  "status":"passing"
}

{
  "subject":"coverage",
  "color":"red",
  "status":"21%"
}
```

## PUT /badges/< owner >/< project >/< name >
Update an existing badge or create a new one if one doesn't exist

### JSON data
* subject - (defaults to `name`) What you want to track the value/status for
* color - The color of the badge
* status - The value/status of the subject

Visit https://shields.io towards the bottom at **Your Badge** for an idea of configuration options.

Sample post data:
```
{
  "color":"green",
  "status":"passing"
}

{
  "subject":"coverage",
  "color":"red",
  "status":"21%"
}
```

## DELETE /badges/< owner >/< project >/< name >
Delete a badge

# Contributing
Contributions, questions, and comments are welcomed and encouraged.

Please open an [Issue](https://github.com/jmccann/badges/issues/new) with any problems you run into or questions you may have.

Please open a [Pull Request](https://github.com/jmccann/badges/compare) with any fixes/enhancements you'd like to contribute.

# Testing
Run locally by [installing Docker](https://www.docker.com/products/overview#/install_the_platform) and the [Drone CLI](http://readme.drone.io/0.5/reference/cli/overview/).  Then run:

```
drone exec
```

This will run the testing inside a docker container.


CHANGELOG.md
## 0.1.0
- Initial Commit


spec/spec_helper.rb
require 'pry'
require 'sinatra'
require 'rack/test'
require 'capybara/rspec'
require 'simplecov'
SimpleCov.start do
  add_filter '.bundler/'
  add_filter 'spec/'
end

# If RACK_ENV isn't set, set it to 'test'.  Sinatra defaults to development,
# so we have to override that unless we want to set RACK_ENV=test from the
# command line when we run rake spec.  That's tedious, so do it here.
ENV['RACK_ENV'] ||= 'test'
set :environment, :test

Dir['lib/**/*.rb'].each do |file|
  require ::File.expand_path("../../#{file.gsub('.rb', '')}", __FILE__)
end

def test_svg # rubocop:disable MethodLength
  '<svg xmlns="http://www.w3.org/2000/svg" width="66" height="20">' \
  '<linearGradient id="b" x2="0" y2="100%">' \
  '<stop offset="0" stop-color="#bbb" stop-opacity=".1"/>' \
  '<stop offset="1" stop-opacity=".1"/></linearGradient><mask id="a">' \
  '<rect width="66" height="20" rx="3" fill="#fff"/></mask><g mask="url(#a)">' \
  '<path fill="#555" d="M0 0h31v20H0z"/>' \
  '<path fill="#97CA00" d="M31 0h35v20H31z"/>' \
  '<path fill="url(#b)" d="M0 0h66v20H0z"/></g>' \
  '<g fill="#fff" text-anchor="middle" ' \
  'font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11">' \
  '<text x="15.5" y="15" fill="#010101" fill-opacity=".3">test</text>' \
  '<text x="15.5" y="14">test</text>' \
  '<text x="47.5" y="15" fill="#010101" fill-opacity=".3">97%</text>' \
  '<text x="47.5" y="14">97%</text></g></svg>'
end

RSpec.configure do |config|
  config.include Rack::Test::Methods

  config.before(:each) do
    Badge.auto_migrate!
    Badge.new(owner: 'jmccann', project: 'app1', image: test_svg,
              name: 'coverage', full_name: 'jmccann/app1/coverage',
              created_at: DateTime.now).save
    Badge.new(owner: 'jmccann', project: 'app1', image: test_svg,
              name: 'climate', full_name: 'jmccann/app1/climate',
              created_at: DateTime.now).save
    Badge.new(owner: 'jmccann', project: 'app2', image: test_svg,
              name: 'coverage', full_name: 'jmccann/app2/coverage',
              created_at: DateTime.now).save
    Badge.new(owner: 'jdoe', project: 'app1', image: test_svg,
              name: 'coverage', full_name: 'jdoe/app1/coverage',
              created_at: DateTime.now).save
  end
end

def app
  Sinatra::Application
end

Capybara.app = app


spec/badges/get_spec.rb
require 'spec_helper'

describe 'get /badges' do
  it 'should list badges' do
    get '/badges'
    expect(last_response).to be_ok
    expect(JSON.parse(last_response.body).count).to eq 4
  end

  it 'should list each badge by their full name' do
    get '/badges'
    expect(JSON.parse(last_response.body).first).to eq 'jmccann/app1/coverage'
  end
end

describe 'get /badges/:owner' do
  it 'should return all badges for an owner' do
    get '/badges/jmccann'

    expect(last_response).to be_ok
    expect(JSON.parse(last_response.body).count).to eq 3
  end
end

describe 'get /badges/:owner/:project/:name' do
  it 'should return 1 badge' do
    get '/badges/jmccann/app1/climate'

    expect(last_response).to be_ok
    expect(JSON.parse(last_response.body)).to be_a Hash
  end

  it 'should return the badge details' do
    get '/badges/jmccann/app1/climate'

    expect(JSON.parse(last_response.body))
      .to include('id', 'created_at', 'full_name',
                  'owner' => 'jmccann', 'project' => 'app1',
                  'name' => 'climate')
  end
end

describe 'get /badges/:owner/:project/:name/badge.svg' do
  it 'should return 1 badge' do
    get '/badges/jmccann/app1/climate/badge.svg'

    expect(last_response).to be_ok
    expect(last_response.body).to be_a String
  end

  it 'should return with svg content_type' do
    get '/badges/jmccann/app1/climate/badge.svg'

    expect(last_response).to be_ok
    expect(last_response.content_type).to eq 'image/svg+xml'
  end

  it 'should return the badge details' do
    get '/badges/jmccann/app1/climate/badge.svg'

    expect(last_response.body).to match(/<svg/)
  end
end


spec/badges/put_spec.rb
require 'spec_helper'

describe 'put /badges/:owner/:project/:name' do
  let(:headers) do
    { 'Content-Type' => 'application/json' }
  end

  it 'updates the badge' do
    badge = Badge.all(owner: 'jmccann', project: 'app1', name: 'climate').first
    expect(badge.image).to match(/>97%</)

    put '/badges/jmccann/app1/climate', {
      color: 'red',
      status: '27%25'
    }.to_json, headers

    # Make sure we are redirecting
    expect(last_response.status).to eq 302
    follow_redirect!

    expect(JSON.parse(last_response.body)['image']).not_to match(/>97%</)
    expect(JSON.parse(last_response.body)['image']).to match(/>27%</)
    badge = Badge.all(owner: 'jmccann', project: 'app1', name: 'climate').first
    expect(badge.image).to match(/>27%</)
  end

  it 'returns updated details of the badge' do
    put '/badges/jmccann/app1/climate', {
      color: 'red',
      status: '27%25'
    }.to_json, headers

    # Make sure we are redirecting
    expect(last_response.status).to eq 302
    expect(last_response).to be_redirect
    follow_redirect!

    expect(JSON.parse(last_response.body))
      .to include('id', 'created_at', 'full_name',
                  'owner' => 'jmccann', 'project' => 'app1',
                  'name' => 'climate')
  end

  it 'creates badge if it doesn not exist' do
    expect(Badge.count).to eq 4
    expect(Badge).to receive(:new).and_call_original

    put '/badges/jmccann/app1/test', {
      color: 'green',
      status: '97%25'
    }.to_json, headers

    # Make sure we are redirecting
    expect(last_response.status).to eq 302
    follow_redirect!

    # Make sure we get back object details
    expect(JSON.parse(last_response.body))
      .to include('id', 'created_at', 'image',
                  'owner' => 'jmccann', 'project' => 'app1', 'name' => 'test',
                  'full_name' => 'jmccann/app1/test')

    # Make sure DB count increased by 1
    expect(Badge.count).to eq 5
  end
end


lib/badges/get.rb
require 'sinatra'
require 'json'

require_relative '../helpers/datamapper'

#
# Show all badges
#
get '/badges' do
  badges = Badge.all
  [200, badges.map { |b| "#{b.owner}/#{b.project}/#{b.name}" }.to_json]
end

#
# Show all badges for an owner
#
get '/badges/:owner' do |owner|
  badges = Badge.all(owner: owner)
  [200, badges.to_json]
end

#
# Show all badges for an owner/project
#
# TBD
#

#
# Show badge details
#
get '/badges/:owner/:project/:name' do |owner, project, name|
  badge = Badge.all(owner: owner, project: project, name: name)
  return [200, badge.first.to_json] if badge.count == 1
  [500, message: "Too many badges returned: #{badge.count}"] if badge.count > 1
end

#
# Return badge
#
get '/badges/:owner/:project/:name/badge.svg' do |owner, project, name|
  content_type 'image/svg+xml'
  badge = Badge.all(owner: owner, project: project, name: name)
  return [200, badge.first.image] if badge.count == 1
  [500, message: "Too many badges returned: #{badge.count}"] if badge.count > 1
end


lib/badges/put.rb
require 'sinatra'
require 'json'
require 'uri'

require_relative '../helpers/datamapper'
require_relative '../helpers/data'
require_relative '../helpers/svg'

put '/badges/:owner/:project/:name' do |owner, project, name|
  subject = data.key?('subject') ? data['subject'] : name

  badges = Badge.all(owner: owner, project: project, name: name)
  if badges.count > 1
    return [500, message: "Expected 1 badge, got #{badges.count}"]
  end

  badge = nil
  if badges.count.zero?
    badge = new_badge(owner, project, name, data)
  else
    badge = badges.first
    badge.image = svg(subject, URI.decode(data['status']), data['color'])
  end

  # After creating object return details of it
  redirect "/badges/#{badge.full_name}" if badge.save

  error = badge.errors.full_messages.join(', ')
  return [500, { message: "Failed to update the Badge: #{error}" }.to_json]
end


lib/helpers/svg.rb
require 'rasem'
require 'rmagick'

def get_hex_color(color, value)
  return color if color =~ /#/
  return color_from_range value.to_i if value =~ /\A[-+]?\d+\z/
  return method(color).call if respond_to?(color, :include_private)
  raise 'Could not determine color'
end

def color_from_range(int) # rubocop:disable MethodLength
  case int
  when 95..100
    brightgreen
  when 90..94
    green
  when 85..89
    yellowgreen
  when 80..84
    yellow
  when 70..79
    orange
  else
    red
  end
end

def svg(subject, value, color = 'green') # rubocop:disable AbcSize, MethodLength
  # color = respond_to?(color, :include_private) ? method(color).call : color
  color = get_hex_color(color, value)

  subject_width = text_width subject
  value_width = text_width value
  width = subject_width + value_width + 20

  img = Rasem::SVGImage.new(width: width, height: 20) do
    l1 = linearGradient('b', x2: '0', y2: '100%') do
      stop('0', '#bbb', '.1')
      stop('1', '#000', '.1')
    end

    mask(id: 'a') do
      rect(0, 0, width, 20, 3, fill: '#fff')
    end

    group(mask: 'url(#a)') do
      path(fill: '#555') do
        moveToA(0, 0)
        hlineTo(subject_width + 10)
        vlineTo(20)
        hlineToA(0)
        close
      end
      path(fill: color) do
        moveToA(subject_width + 10, 0)
        hlineTo(value_width + 10)
        vlineTo(20)
        hlineToA(subject_width + 10)
        close
      end
      path(fill: l1.fill) do
        moveToA(0, 0)
        hlineTo(width)
        vlineTo(20)
        hlineToA(0)
        close
      end
    end

    group(fill: '#fff', 'text-anchor' => 'middle',
          'font-family' => 'DejaVu Sans,Verdana,Geneva,sans-serif',
          'font-size' => 11) do
      text((subject_width / 2) + 5, 15,
           fill: '#010101', 'fill-opacity' => '.3') do
        raw subject
      end
      text((subject_width / 2) + 5, 14, fill: '#fff') do
        raw subject
      end
      text(subject_width + (value_width / 2) + 15, 15,
           fill: '#010101', 'fill-opacity' => '.3') do
        raw value
      end
      text(subject_width + (value_width / 2) + 15, 14, fill: '#fff') do
        raw value
      end
    end
  end
  img.to_s
end

def brightgreen
  '#4C1'
end

def green
  '#97CA00'
end

def yellowgreen
  '#A4A61D'
end

def yellow
  '#DFB317'
end

def orange
  '#FE7D37'
end

def red
  '#E05D44'
end

#
# Determine text width rounded up to next multiple of 10
#
def text_width(text)
  label = Magick::Draw.new
  label.font = 'DejaVu Sans'
  label.text_antialias true
  label.font_size 11
  label.font_style = Magick::NormalStyle
  label.gravity = Magick::CenterGravity
  label.text(0, 0, text)
  metrics = label.get_type_metrics(text)
  metrics.width.to_i.round(-1)
end
